### PR TITLE
Initial state, validations, form submission.

### LINKED ISSUE #1: Support initial state
The `Form` class accepts an `initialState` argument, at the moment this doesn't do anything however it should convert a vanilla object into Maps and Stacks to fit the immutable structure of a Form. This can be done using the `reviver` callback passed to `fromJS` but needs further investigation.

### PR DESCRIPTION
**Pull request by TimMikeladze**

### CODE REVIEW

**Comment by TimMikeladze** [comment]
@alexyuly Please review

File: src/Form.js [hunk 1]
```diff
@@ -56,7 +61,7 @@ const createReducer = initialState =>
       }
       case ADD_ERROR: {
         const { error } = action.payload;
-        const errorsStack = state.get('errors').push(error);
+        const errorsStack = state.has('errors') ? state.get('errors').push(error) : Stack().push(error);
```
Comment (alexyuly) [inline src/Form.js]: More proper to write `Stack.of(error)` than `Stack().push(error)`. See above too.

File: src/Form.js [hunk 2]
```diff
@@ -71,31 +76,93 @@ const createReducer = initialState =>
...
   };
 
+const reviver = (key, value) => {
+  let iterable;
+  if (key === 'errors') {
+    iterable = value.toStack();
+  } else {
+    iterable = value.toMap();
+  }
+  return iterable;
```
Comment (alexyuly) [inline src/Form.js]: Write this as a ternary statement?

    const reviver = (key, value) => (key === 'errors' ? value.toStack() : value.toMap())

File: src/Form.js [hunk 3]
```diff
@@ -71,31 +76,93 @@ const createReducer = initialState =>
...
+};
+
+export { reviver };
+
+const filterValidate = (state) => {
+  let fields = {};
+  const form = Object.assign({}, state);
```
Comment (alexyuly) [inline src/Form.js]: In ES6, I prefer `{ ...state }`

File: src/Form.js [hunk 4]
```diff
@@ -71,31 +76,93 @@ const createReducer = initialState =>
...
+    fields = keys(form.fields)
+        .reduce((res, key) => {
+          if (hasIn(form, ['fields', key, 'validate'])) {
+            fieldValidators[key] = Array.isArray(form.fields[key].validate)
+              ? form.fields[key].validate
+              : [form.fields[key].validate];
+            delete form.fields[key].validate;
+          }
+          return { ...res, [key]: form.fields[key] };
+        }, {});
```
Comment (alexyuly) [inline src/Form.js]: Fix 4-space indentation

File: src/Form.js [hunk 5]
```diff
@@ -141,14 +208,61 @@ class Form {
...
+        }).catch((err) => {
+          if (this.onFailure) {
+            this.onFailure(err, { store });
+          }
+          reject(err);
+        });
+      } else {
+        reject('Validation failed');
+      }
+    });
```
Comment (alexyuly) [inline src/Form.js]: If you wanted to get really fancy, you could write this method with `async`/`await`

    async submit(promise) {
      const store = this.store;
      if (this.validate()) {
        let res;
        try {
          res = await promise;
        }
        catch (error) {
          if (this.onFailure) {
            this.onFailure(err, { store });
          }
          throw err;
        }
        if (this.onSuccess) {
          this.onSuccess(res, { store });
        }
        return res;
      }
      else {
        throw(new Error('Validation failed!'));
      }
    }

or something. Doesn't really matter though.

File: src/Form.js (hunk 2 above)
Comment (TimMikeladze) [inline src/Form.js]: Left it like this on purpose. There will most likely be other keys as the package matures.

File: src/Form.js (hunk 5 above)
Comment (TimMikeladze) [inline src/Form.js]: True. I'll add the babel config and do it.

File: src/Form.js (hunk 3 above)
Comment (TimMikeladze) [inline src/Form.js]: Good call.

File: src/Form.js (hunk 4 above)
Comment (TimMikeladze) [inline src/Form.js]: eslint and atom autoformat to this with no errors.

**Pull request merged**